File: sprint11/KMS/SW/BOJ_1654.py
K, N = map(int, input().split())
L = []
for i in range(K):
    L.append(int(input()))
l = 1
r = max(L)
answer = 1
while l <= r:
    mid = (l + r) // 2
    cnt = 0
    for i in L:
        cnt += i // mid
    if cnt < N:
        r = mid - 1
    else:
        if answer < mid:
            answer = mid
        l = mid + 1
print(answer)

File: sprint11/JMS/SW/BOJ_1654.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	long long K, N, right = 0, mid, left = 1, cnt = 0, result = 0;
	long long arr[10001];

	cin >> K >> N;
	for (int i = 0; i < K; i++) {
		cin >> arr[i];
		right = max(right, arr[i]);
	}

	mid = (left + right) / 2;
	while (left <= right) {
		mid = (left + right) / 2;
		cnt = 0;

		for (int i = 0; i < K; i++)
			cnt += arr[i] / mid;

		if (cnt >= N) {
			left = mid + 1;
			result = result > mid ? result : mid;
		}
		else
			right = mid - 1;
	}

	cout << result;
}
